File: trunk/python/psem2m.base/src/base/multicast_agent.py
#!/usr/bin/env python
#-- Content-Encoding: UTF-8 --
"""
Created on 11 juin 2012

:author: Thomas Calmant
"""

# ------------------------------------------------------------------------------

import logging
import socket
import struct
import threading

# ------------------------------------------------------------------------------

from pelix.ipopo.decorators import ComponentFactory, Requires, Validate, \
    Invalidate, Instantiate

# ------------------------------------------------------------------------------

_logger = logging.getLogger(__name__)

# ------------------------------------------------------------------------------

def make_mreq(family, address):
    """
    Makes a mreq structure object for the given address and socket family.
    
    :param family: A socket family (AF_INET or AF_INET6)
    :param address: A multicast address
    :raise ValueError: Invalid family or address
    """
    if not address:
        raise ValueError("Empty address")

    # Convert the address to a binary form
    group_bin = socket.inet_pton(family, address)

    if family == socket.AF_INET:
        # IPv4
        # struct ip_mreq
        # {
        #     struct in_addr imr_multiaddr; /* IP multicast address of group */
        #     struct in_addr imr_interface; /* local IP address of interface */
        # };
        # "=I" : Native order, standard size unsigned int
        return group_bin + struct.pack("=I", socket.INADDR_ANY)

    elif family == socket.AF_INET6:
        # IPv6
        # struct ipv6_mreq {
        #    struct in6_addr ipv6mr_multiaddr;
        #    unsigned int    ipv6mr_interface;
        # };
        # "@I" : Native order, native size unsigned int
        return group_bin + struct.pack("@I", 0)

    raise ValueError("Unknown family %s", family)


def create_multicast_socket(address, port):
    """
    Creates a multicast socket according to the given address and port.
    Handles both IPv4 and IPv6 addresses.
    
    :raise ValueError: Invalid address or port
    """
    # Get the information about a datagram (UDP) socket, of any family
    try:
        addr_info = socket.getaddrinfo(address, port, socket.AF_UNSPEC,
                                       socket.SOCK_DGRAM)
    except socket.gaierror:
        raise ValueError("Error retrieving address informations (%s, %s)" \
                         % (address, port))

    if len(addr_info) > 1:
        _logger.debug("More than one address information found. "
                      "Using the first one.")

    # Get the first entry : (family, socktype, proto, canonname, sockaddr)
    addr_info = addr_info[0]

    # Only accept IPv4/v6 addresses
    if addr_info[0] not in (socket.AF_INET, socket.AF_INET6):
        # Unhandled address family
        raise ValueError("Unhandled socket family : %d" % (addr_info[0]))

    # Prepare the socket
    sock = socket.socket(addr_info[0], socket.SOCK_DGRAM)

    # Reuse address
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    if hasattr(socket, 'SO_REUSEPORT'):
        # Special for MacOS
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEPORT, 1)

    # Bind the socket
    if sock.family == socket.AF_INET:
        # IPv4 binding
        sock.bind(('0.0.0.0', port))

    else:
        # IPv6 Binding
        sock.bind(('::', port))

    # Prepare the mreq structure to join the group
    # addrinfo[4] = (addr,port)
    mreq = make_mreq(sock.family, addr_info[4][0])

    # Join the group
    if sock.family == socket.AF_INET:
        # IPv4
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_ADD_MEMBERSHIP, mreq)

        # Allow multicast packets to get back on this host
        sock.setsockopt(socket.IPPROTO_IP, socket.IP_MULTICAST_LOOP, 1)

    else:
        # IPv6
        sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_JOIN_GROUP, mreq)

        # Allow multicast packets to get back on this host
        sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_MULTICAST_LOOP, 1)

    return (sock, addr_info[4][0])


def close_multicast_socket(sock, address):
    """
    Cleans up the given multicast socket.
    Unregisters it of the multicast group.
    
    Parameters should be the result of create_multicast_socket
    
    :param sock: A multicast socket
    :param address: The multicast address used by the socket
    """
    if not sock:
        return

    if address:
        # Prepare the mreq structure to join the group
        mreq = make_mreq(sock.family, address)

        # Quit group
        if sock.family == socket.AF_INET:
            # IPv4
            sock.setsockopt(socket.IPPROTO_IP, socket.IP_DROP_MEMBERSHIP, mreq)

        elif sock.family == socket.AF_INET6:
            # IPv6
            sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_LEAVE_GROUP, mreq)

    # Close the socket
    sock.close()

# ------------------------------------------------------------------------------

# FIXME: It should set up by the configuration service
PSEM2M_MULTICAST_PORT = 42000
""" The PSEM2M multicast port """

SIGNAL_PREFIX = "/psem2m-multicast-agent"
SIGNAL_MATCH_ALL = "%s/*" % SIGNAL_PREFIX
SIGNAL_CONFIRM_BEAT = "%s/confirm" % SIGNAL_PREFIX

PACKET_REGISTER = 1
""" Packet type 'Register' : for isolate first registration """

@ComponentFactory("psem2m-multicast-agent-factory")
@Instantiate("psem2m-multicast-agent")
@Requires("_config", "org.psem2m.isolates.services.conf.ISvcConfig")
@Requires("_directory", "org.psem2m.signals.ISignalDirectory")
@Requires("_receiver", "org.psem2m.signals.ISignalReceiver")
@Requires("_sender", "org.psem2m.signals.ISignalBroadcaster")
class MulticastAgent(object):
    """
    An UDP v4/v6 multicast agent for PSEM2M
    """
    def __init__(self):
        """
        Constructor
        """
        self._config = None
        self._directory = None
        self._receiver = None
        self._sender = None

        self._target = None
        self._socket = None

        self._thread = None
        self._thread_running = False


    def handle_received_signal(self, name, data):
        """
        Handles a received signal
        
        :param name: Signal name
        :param data: Complete signal data (meta-data + content)
        :return: Data to send to the requester, or None
        """
        if name == SIGNAL_CONFIRM_BEAT:
            # Beat confirmation
            isolate_id = self._directory.get_isolate_id()

            # FIXME: setup groups from the configuration service
            # FIXME: the virtual group ALL should be computed by the directory
            groups = ["ALL"]

            if isolate_id.startswith("org.psem2m.internals.isolates.forker"):
                # Forkers can only be forkers
                groups.append("FORKERS")

            else:
                # A forker can't be an isolate and vice versa
                groups.append("ISOLATES")

            if isolate_id.startswith("org.psem2m.internals.isolates.monitor"):
                # A monitor is a special isolate
                groups.append("MONITORS")

            return {'id': isolate_id,
                    'node': self._directory.get_local_node(),
                    'groups': groups}


    @Invalidate
    def invalidate(self, context):
        """
        Component invalidated
        
        :param context: The bundle context
        """
        self._thread_running = False

        # Unregister to the signals
        self._receiver.unregister_listener(SIGNAL_MATCH_ALL, self)

        # Close the socket
        close_multicast_socket(self._socket, self._target[0])
        self._target = None
        self._socket = None


    @Validate
    def validate(self, context):
        """
        Component validated
        
        :param context: The bundle context
        """
        # Get the application multicast group
        multicast = self._config.get_application().get_multicast()

        # Create the corresponding socket
        self._socket, address = create_multicast_socket(multicast,
                                                        PSEM2M_MULTICAST_PORT)
        self._target = (address, PSEM2M_MULTICAST_PORT)

        # Start the reading thread
        self._thread_running = True
        self._thread = threading.Thread(target=self._run, name="MulticastAgent")
        self._thread.start()

        # Register to signals
        self._receiver.register_listener(SIGNAL_MATCH_ALL, self)

        # Send the registration packet
        self._send_beat()


    def _handle_registration(self, data, host):
        """
        Handles the registration of an isolate
        """
        try:
            # Little endian
            port = struct.unpack("<H", data[1:])[0]

            # Ask for a confirmation
            response = self._sender.send_to(SIGNAL_CONFIRM_BEAT, None,
                                            host, port)

            if response and response.get('results', None):
                # We got at least one result.
                # Register the isolate, considering the first result only
                results = response['results'][0]

                isolate_id = results['id']
                node = results['node']
                groups = results['groups']

                if self._directory is not None:
                    # Directory can be None in the end of the instance
                    self._directory.register_isolate(isolate_id, node,
                                                     host, port, groups)

            else:
                _logger.error("The isolate at [%s]:%d didn't confirmed "
                              "its existence", host, port)

        except:
            _logger.exception("Error confirming an agent beat "
                              "from [%s]:%d", host, port)

    def _run(self):
        """
        Thread listening to the multicast socket
        """
        while self._thread_running:
            try:
                _logger.debug("WAITING FOR DATA....")
                data, sender = self._socket.recvfrom(1500)
                _logger.debug("GOT DATA from '%s' - '%r'", sender, data)

                # Get the content
                if data[0] == PACKET_REGISTER:
                    # Send a signal to (sender[0], content["port"])
                    host = sender[0]

                    # Register in a new thread
                    threading.Thread(target=self._handle_registration,
                                     args=(data, host)).start()

            except:
                if self._thread_running:
                    # Only log if it's a "live" error
                    _logger.exception("Error reading multicast socket")


    def _send_beat(self):
        """
        Forges and sends the registration packet
        """
        port = self._receiver.get_access_info()[1]

        # Prepare the packet
        packet = struct.pack("<BH", PACKET_REGISTER, port)

        # Send over the network
        self._socket.sendto(packet, 0, self._target)

File: org.psem2m.signals.multicast/src/org/psem2m/signals/multicast/MulticastAgent.java
/**
 * File:   MulticastAgent.java
 * Author: Thomas Calmant
 * Date:   14 juin 2012
 */
package org.psem2m.signals.multicast;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.InetAddress;
import java.net.MulticastSocket;
import java.net.UnknownHostException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.apache.felix.ipojo.annotations.Component;
import org.apache.felix.ipojo.annotations.Instantiate;
import org.apache.felix.ipojo.annotations.Requires;
import org.apache.felix.ipojo.annotations.Validate;
import org.psem2m.isolates.base.IIsolateLoggerSvc;
import org.psem2m.isolates.constants.IPlatformProperties;
import org.psem2m.isolates.services.conf.ISvcConfig;
import org.psem2m.signals.HostAccess;
import org.psem2m.signals.ISignalBroadcaster;
import org.psem2m.signals.ISignalData;
import org.psem2m.signals.ISignalDirectory;
import org.psem2m.signals.ISignalListener;
import org.psem2m.signals.ISignalReceiver;

/**
 * Implementation of the multicast agent : sends a packet when validated and
 * completes the signals directory when it receives a packet.
 * 
 * @author Thomas Calmant
 */
@Component(name = "psem2m-multicast-agent-factory", publicFactory = false)
@Instantiate(name = "psem2m-multicast-agent")
public class MulticastAgent implements ISignalListener {

    /**
     * Leaves the group and closes the multicast socket.
     * 
     * @param aSocket
     *            A multicast socket
     * @param aAddress
     *            The multicast address
     * @throws IOException
     *             Error reading the address or leaving the group
     */
    public static void closeMulticast(final MulticastSocket aSocket,
            final InetAddress aAddress) throws IOException {

        if (aSocket == null) {
            // Nothing to do
            return;
        }

        try {
            // Leave the group
            aSocket.leaveGroup(aAddress);

        } finally {
            // Close the socket
            aSocket.close();
        }
    }

    /**
     * Sets up a multicast socket
     * 
     * @param aAddress
     *            The multicast address (group)
     * @param aPort
     *            The multicast port
     * @return The created socket
     * @throws IOException
     *             Something wrong occurred (bad address, bad port, ...)
     */
    public static MulticastSocket setupMulticast(final InetAddress aAddress,
            final int aPort) throws IOException {

        // Set up the socket
        final MulticastSocket socket = new MulticastSocket(aPort);
        socket.setLoopbackMode(true);
        socket.setReuseAddress(true);

        // Join the group
        try {
            socket.joinGroup(aAddress);

        } catch (final IOException ex) {
            // Be nice...
            socket.close();
            throw ex;
        }

        return socket;
    }

    /** The configuration service */
    @Requires
    private ISvcConfig pConfig;

    /** The signals directory */
    @Requires
    private ISignalDirectory pDirectory;

    /** The logger */
    @Requires
    private IIsolateLoggerSvc pLogger;

    /** The multicast address */
    private InetAddress pMulticastGroup;

    /** The multicast port */
    private int pMulticastPort;

    /** The signals receiver */
    @Requires
    private ISignalReceiver pReceiver;

    /** The signals sender */
    @Requires
    private ISignalBroadcaster pSender;

    /** The socket */
    private MulticastSocket pSocket;

    /** The listening thread */
    private Thread pThread;

    /** The thread loop control */
    private boolean pThreadRun;

    /*
     * (non-Javadoc)
     * 
     * @see
     * org.psem2m.signals.ISignalListener#handleReceivedSignal(java.lang.String,
     * org.psem2m.signals.ISignalData)
     */
    @Override
    public Object handleReceivedSignal(final String aSignalName,
            final ISignalData aSignalData) {

        if (IMulticastConstants.SIGNAL_CONFIRM_BEAT.equals(aSignalName)) {
            // Beat confirmation
            final String isolateId = pDirectory.getIsolateId();

            // Compute the groups
            // FIXME: this should be done in the configuration service
            final List<String> groups = new ArrayList<String>();

            // FIXME: this virtual group should be computed by the directory
            groups.add("ALL");

            if (isolateId
                    .startsWith(IPlatformProperties.SPECIAL_ISOLATE_ID_FORKER)) {
                // ... we are a forker (just in case)
                groups.add("FORKERS");

            } else {
                // ... we are an isolate (only the forkers are out of this
                // group)
                groups.add("ISOLATES");
            }

            if (isolateId
                    .startsWith(IPlatformProperties.SPECIAL_ISOLATE_ID_MONITOR)) {
                // ... we are a monitor
                groups.add("MONITORS");
            }

            // Return the result map
            final Map<String, Object> result = new HashMap<String, Object>();
            result.put("id", isolateId);
            result.put("node", pDirectory.getLocalNode());
            result.put("groups", groups);
            return result;
        }

        // Unknown service
        return null;
    }

    /**
     * Decodes a registration packet and registers the isolate if it responds to
     * a confirmation signal
     * 
     * @param aData
     *            Registration packet content
     * @param aHostAddress
     *            The sender address
     */
    protected void handleRegistration(final byte[] aData,
            final String aHostAddress) {

        // Make a little endian byte array reader
        final ByteBuffer buffer = ByteBuffer.wrap(aData, 1, aData.length - 1);
        buffer.order(ByteOrder.LITTLE_ENDIAN);

        // Get the port
        final int port = buffer.getShort();

        // Send the confirmation signal
        final Object[] results;
        try {
            results = pSender.sendTo(IMulticastConstants.SIGNAL_CONFIRM_BEAT,
                    null, aHostAddress, port);

        } catch (final Exception e) {
            // Error sending the signal, stop here
            pLogger.logWarn(this, "handleRegistration",
                    "Error sending a confirmation beat to host=", aHostAddress,
                    "port=", port, "ex=", e);
            return;
        }

        // We need at least one result
        if (results == null || results.length == 0) {
            pLogger.logWarn(
                    this,
                    "handleRegistration",
                    "Empty beat confirmation : not enough data to register the isolate. host=",
                    aHostAddress, "port=", port);
            return;
        }

        // Find the first map in the result
        Map<?, ?> map = null;
        for (final Object result : results) {
            if (result instanceof Map) {
                map = (Map<?, ?>) result;
                break;
            }
        }

        if (map == null) {
            pLogger.logWarn(this, "handleRegistration",
                    "No readable result for host=", aHostAddress, "port=", port);
            return;
        }

        // Get the response content
        final String isolate = (String) map.get("id");
        final String node = (String) map.get("node");
        final Object rawGroups = map.get("groups");

        if (isolate == null || isolate.isEmpty() || node == null
                || node.isEmpty() || rawGroups == null) {
            // Invalid content
            pLogger.logSevere(this, "handleRegistration",
                    "Invalid content for host=", aHostAddress, "port=", port,
                    "isolate=", isolate, "node=", node, "rawGroups=", rawGroups);
            return;
        }

        // Convert the raw groups
        final String[] groups;
        if (rawGroups instanceof Collection) {
            // Collection found, convert it
            groups = ((Collection<?>) rawGroups).toArray(new String[0]);

        } else if (rawGroups instanceof String[]) {
            // Strings array found, use it directly
            groups = (String[]) rawGroups;

        } else if (rawGroups instanceof Object[]) {
            // Objects array found, convert it
            final List<String> list = new ArrayList<String>();
            for (final Object object : (Object[]) rawGroups) {
                if (object != null) {
                    list.add((String) object);
                }
            }
            groups = list.toArray(new String[0]);

        } else {
            // Unknown type
            pLogger.logWarn(this, "handleRegistration",
                    "Unknown groups container type=", rawGroups.getClass()
                            .getName());
            return;
        }

        // Update the address associated to the node
        pDirectory.setNodeAddress(node, aHostAddress);

        // Register the isolate
        pDirectory.registerIsolate(isolate, node, port, groups);
    }

    /**
     * Component invalidated
     */
    public void invalidate() {

        // Unregister from signals receiver
        pReceiver
                .unregisterListener(IMulticastConstants.SIGNAL_MATCH_ALL, this);

        // Stop the thread...
        pThreadRun = false;

        // ... the hard way
        pThread.interrupt();

        // Close the socket
        try {
            closeMulticast(pSocket, pMulticastGroup);

        } catch (final IOException e) {
            pLogger.logSevere(this, "invalidate",
                    "Error closing the multicast socket, ex=", e);
        }
    }

    /**
     * Waits for packets on the multicast socket
     */
    protected void receivePackets() {

        // Set up the buffer
        final byte[] buffer = new byte[1500];
        final DatagramPacket packet = new DatagramPacket(buffer, buffer.length);

        while (pThreadRun) {

            // Clear the buffer
            Arrays.fill(buffer, (byte) 0);

            try {
                // Wait for a packet (blocking)
                pSocket.receive(packet);

                // Get the content
                final byte[] data = packet.getData();

                pLogger.logInfo(this, "receivePackets", "Got packet, data=",
                        Arrays.toString(data));

                switch (data[0]) {
                case IMulticastConstants.PACKET_REGISTER:
                    // Register the isolate in another thread
                    new Thread(new Runnable() {

                        @Override
                        public void run() {

                            handleRegistration(data, packet.getAddress()
                                    .getHostAddress());
                        }
                    }).start();
                    break;
                }

            } catch (final IOException ex) {
                // Just log
                pLogger.logWarn(this, "receivePackets",
                        "Error receiving a packet, ex=", ex);
            }
        }
    }

    /**
     * Sends the beat
     * 
     * @throws IOException
     *             Error sending the beat
     */
    protected void sendBeat() throws IOException {

        // Get the access port
        final HostAccess access = pReceiver.getAccessInfo();
        if (access == null) {
            pLogger.logSevere(this, "sendBeat",
                    "This isolate is not accessible");
            return;
        }

        // Prepare the content : 3 bytes (type, port)
        final ByteBuffer buffer = ByteBuffer.allocate(3);
        buffer.order(ByteOrder.LITTLE_ENDIAN);

        // ... packet type
        buffer.put(IMulticastConstants.PACKET_REGISTER);

        // ... access port
        buffer.putShort((short) access.getPort());

        // Prepare the packet
        buffer.position(0);
        final byte[] content = new byte[buffer.remaining()];
        buffer.get(content);

        final DatagramPacket packet = new DatagramPacket(content,
                content.length, pMulticastGroup, pMulticastPort);

        // Send it
        pSocket.send(packet);

        pLogger.logDebug(this, "sendBeat", "Sent beat=",
                Arrays.toString(content), "to=", pMulticastGroup, "port=",
                pMulticastPort);
    }

    /**
     * Component validated
     */
    @Validate
    public void validate() {

        // Register to the signals
        pReceiver.registerListener(IMulticastConstants.SIGNAL_MATCH_ALL, this);

        // Get the multicast group and port
        try {
            pMulticastGroup = InetAddress.getByName(pConfig.getApplication()
                    .getMulticast());

            pMulticastPort = IMulticastConstants.PSEM2M_MULTICAST_PORT;

        } catch (final UnknownHostException ex) {
            pLogger.logSevere(this, "validate",
                    "Couldn't read the multicast group=", pConfig
                            .getApplication().getMulticast());
            return;
        }

        try {
            // Create the socket (can fail)
            pSocket = setupMulticast(pMulticastGroup, pMulticastPort);

            // Socket created, start the thread
            pThreadRun = true;
            pThread = new Thread(new Runnable() {

                @Override
                public void run() {

                    receivePackets();
                }
            }, "Multicast-Agent");
            pThread.start();

        } catch (final IOException e) {
            pLogger.logSevere(this, "validate",
                    "Couldn't start the multicast socket - group=",
                    pMulticastGroup, "ex=", e);
            return;
        }

        try {
            // Send the beat
            sendBeat();
            pLogger.logInfo(this, "validate", "Multicast agent ready");

        } catch (final IOException e) {
            pLogger.logSevere(this, "validate",
                    "Error sending the registration beat, ex=", e);
        }
    }
}
